### ISSUE TITLE
focal_loss = False

### ISSUE THREAD

**Issue by wangpupanjing** [supernotman/RetinaFace_Pytorch#8]
focal_loss = False
            # focal loss
            if focal_loss:
                alpha = 0.25
                gamma = 2.0            
                alpha_factor = torch.ones(targets.shape).cuda() * alpha

                alpha_factor = torch.where(torch.eq(targets, 1.), alpha_factor, 1. - alpha_factor)
                focal_weight = torch.where(torch.eq(targets, 1.), 1. - classification, classification)
                focal_weight = alpha_factor * torch.pow(focal_weight, gamma)

                bce = -(targets * torch.log(classification) + (1.0 - targets) * torch.log(1.0 - classification))

                cls_loss = focal_weight * bce

                cls_loss = torch.where(torch.ne(targets, -1.0), cls_loss, torch.zeros(cls_loss.shape).cuda())

                classification_losses.append(cls_loss.sum()/torch.clamp(num_positive_anchors.float(), min=1.0))
            else:
                if positive_indices.sum() > 0:
                    classification_losses.append(positive_losses.mean() + sorted_losses.mean())
                else:
                    classification_losses.append(torch.tensor(0).float().cuda())

never use focalloss???

**Comment by luyao777** [comment]
Maybe there is no need to use focal loss

**Comment by supernotman** [comment]
> focal_loss = False
> # focal loss
> if focal_loss:
> alpha = 0.25
> gamma = 2.0
> alpha_factor = torch.ones(targets.shape).cuda() * alpha
> 
> ```
>             alpha_factor = torch.where(torch.eq(targets, 1.), alpha_factor, 1. - alpha_factor)
>             focal_weight = torch.where(torch.eq(targets, 1.), 1. - classification, classification)
>             focal_weight = alpha_factor * torch.pow(focal_weight, gamma)
> 
>             bce = -(targets * torch.log(classification) + (1.0 - targets) * torch.log(1.0 - classification))
> 
>             cls_loss = focal_weight * bce
> 
>             cls_loss = torch.where(torch.ne(targets, -1.0), cls_loss, torch.zeros(cls_loss.shape).cuda())
> 
>             classification_losses.append(cls_loss.sum()/torch.clamp(num_positive_anchors.float(), min=1.0))
>         else:
>             if positive_indices.sum() > 0:
>                 classification_losses.append(positive_losses.mean() + sorted_losses.mean())
>             else:
>                 classification_losses.append(torch.tensor(0).float().cuda())
> ```
> 
> never use focalloss???

I tried focalloss but failed.The model even can't converge while training.

**Comment by supernotman** [comment]
> > focal_loss = False
> > # focal loss
> > if focal_loss:
> > alpha = 0.25
> > gamma = 2.0
> > alpha_factor = torch.ones(targets.shape).cuda() * alpha
> > ```
> >             alpha_factor = torch.where(torch.eq(targets, 1.), alpha_factor, 1. - alpha_factor)
> >             focal_weight = torch.where(torch.eq(targets, 1.), 1. - classification, classification)
> >             focal_weight = alpha_factor * torch.pow(focal_weight, gamma)
> > 
> >             bce = -(targets * torch.log(classification) + (1.0 - targets) * torch.log(1.0 - classification))
> > 
> >             cls_loss = focal_weight * bce
> > 
> >             cls_loss = torch.where(torch.ne(targets, -1.0), cls_loss, torch.zeros(cls_loss.shape).cuda())
> > 
> >             classification_losses.append(cls_loss.sum()/torch.clamp(num_positive_anchors.float(), min=1.0))
> >         else:
> >             if positive_indices.sum() > 0:
> >                 classification_losses.append(positive_losses.mean() + sorted_losses.mean())
> >             else:
> >                 classification_losses.append(torch.tensor(0).float().cuda())
> > ```
> > 
> > 
> > never use focalloss???
> 
> I tried focalloss but failed.The model even can't converge while training.

I guess some special initialization for model are needed like in retinaNet.(just a guess)
You can have a try of focalloss,maybe i got something wrong.

**Comment by wangpupanjing** [comment]
hi , you even not use  crossentrpy, only .mean+.mean, why ?
is this work ?

**Comment by supernotman** [comment]
> hi , you even not use crossentrpy, only .mean+.mean, why ?
> is this work ?

Actually,i used binary crossentropy loss  in classification.In Pytorch, the ce loss function nn.CrossEntropy actually is a combination of nn.logsoftmax and nn.nllloss().
In my code, i just use nn.logsoftmax() and achieve nn.nllloss() in another way.
About the loss details , a blog for reference:[https://blog.csdn.net/watermelon1123/article/details/91044856](url)

**Comment by wangpupanjing** [comment]
understand, 3Q,  author ,you are very smart, .  can you upload your pretrained model???
